File: rd-net/RdFramework/Impl/RdList.cs
#nullable disable

using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.ComponentModel;
using System.Linq;
using JetBrains.Annotations;
using JetBrains.Collections.Synchronized;
using JetBrains.Collections.Viewable;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.Rd.Base;
using JetBrains.Rd.Util;
using JetBrains.Serialization;

// ReSharper disable InconsistentNaming

namespace JetBrains.Rd.Impl
{
  public class  RdList<V> : RdReactiveBase, IViewableList<V>
    , INotifyCollectionChanged
    where V : notnull
  {
    private readonly ViewableList<V> myList = new(new SynchronizedList<V>()/*to have thread safe print*/);

    public RdList(CtxReadDelegate<V> readValue, CtxWriteDelegate<V> writeValue, long nextVersion = 1L)
    {
      myNextVersion = nextVersion;
      ValueCanBeNull = false;

      ReadValueDelegate = readValue;
      WriteValueDelegate = writeValue;

      //WPF integration
      this.AdviseAddRemove(Lifetime.Eternal, (kind, idx, v) =>
      {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs("Item[]"));
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs("Count"));
        CollectionChanged?.Invoke(this, new NotifyCollectionChangedEventArgs(kind == AddRemove.Add ? NotifyCollectionChangedAction.Add : NotifyCollectionChangedAction.Remove, v, idx));
      });
    }

    //WPF integration
    public event NotifyCollectionChangedEventHandler CollectionChanged;
    public override event PropertyChangedEventHandler PropertyChanged;



    #region Serializers


    [PublicAPI]
    public CtxReadDelegate<V> ReadValueDelegate { get; private set; }
    [PublicAPI]
    public CtxWriteDelegate<V> WriteValueDelegate { get; private set; }

    [PublicAPI]
    public static RdList<V> Read(SerializationCtx ctx, UnsafeReader reader)
    {
      return Read(ctx, reader, Polymorphic<V>.Read, Polymorphic<V>.Write);
    }
    [PublicAPI]
    public static RdList<V> Read(SerializationCtx ctx, UnsafeReader reader,CtxReadDelegate<V> readValue, CtxWriteDelegate<V> writeValue)
    {
      var nextVersion = reader.ReadLong();
      var id = reader.ReadRdId();
      return new RdList<V>(readValue, writeValue, nextVersion).WithId(id);
    }

    [PublicAPI]
    public static void Write(SerializationCtx ctx, UnsafeWriter writer, RdList<V> value)
    {
      Assertion.Assert(!value.RdId.IsNil);
      writer.WriteInt64(value.myNextVersion);
      writer.Write(value.RdId);
    }
    #endregion


    #region Versions

    private const int versionedFlagShift = 2; //change when you change AddUpdateRemove

    private long myNextVersion;
    #endregion


    #region Init


    public bool OptimizeNested { [PublicAPI] get; set; }
    [ItemCanBeNull] private volatile SynchronizedList<LifetimeDefinition> myBindDefinitions;

    protected override void Unbind()
    {
      base.Unbind();
      myBindDefinitions = null;
    }

    protected override void PreInit(Lifetime lifetime, IProtocol proto)
    {
      base.PreInit(lifetime, proto);

      if (!OptimizeNested)
      {
        var definitions = new SynchronizedList<LifetimeDefinition>(null, myList.Count);

        for (var index = 0; index < Count; index++)
        {
          var item = this[index];
          if (item != null)
          {
            item.IdentifyPolymorphic(proto.Identities, proto.Identities.Next(RdId));
            definitions.Add(TryPreBindValue(lifetime, item, index, false));
          }
        }

        using var cookie = lifetime.UsingExecuteIfAlive();
        if (cookie.Succeed)
        {
          Assertion.Assert(myBindDefinitions == null);
          myBindDefinitions = definitions;
        }
        else
          return;
      }

      proto.Wire.Advise(lifetime, this);
    }

    protected override void Init(Lifetime lifetime, IProtocol proto, SerializationCtx ctx)
    {
      base.Init(lifetime, proto, ctx);

      if (!OptimizeNested)
      {
        Change.Advise(lifetime, it =>
        {
          if (IsLocalChange)
          {
            var definitions = TryGetBindDefinitions(lifetime);
            if (definitions == null)
              return;

            if (it.Kind != AddUpdateRemove.Add)
              definitions[it.Index]?.Terminate();

            if (it.Kind == AddUpdateRemove.Remove)
              definitions.RemoveAt(it.Index);

            if (it.Kind != AddUpdateRemove.Remove && it.NewValue != null)
            {
              it.NewValue.IdentifyPolymorphic(proto.Identities, proto.Identities.Next(RdId));
              definitions.Insert(it.Index, TryPreBindValue(lifetime, it.NewValue, it.Index, false));
            }
          }
        });
      }

      using (UsingLocalChange())
      {
        Advise(lifetime, it =>
        {
          if (!IsLocalChange) return;

          proto.Wire.Send(RdId, SendContext.Of(ctx, it, this), static(sendContext, stream) =>
          {
            var sContext = sendContext.SzrCtx;
            var evt = sendContext.Event;
            var me = sendContext.This;

            stream.WriteInt64((me.myNextVersion++ << versionedFlagShift) | (long)evt.Kind);

            stream.WriteInt32(evt.Index);
            if (evt.Kind != AddUpdateRemove.Remove)
              me.WriteValueDelegate(sContext, stream, evt.NewValue);

            SendTrace?.Log($"list `{me.Location}` ({me.RdId}) :: {evt.Kind} :: index={evt.Index} :: " +
                           $"version = {me.myNextVersion - 1}" +
                           $"{(evt.Kind != AddUpdateRemove.Remove ? " :: value = " + evt.NewValue.PrintToString() : "")}");
          });

          if (!OptimizeNested)
            it.NewValue.BindPolymorphic();
        });
      }


    }

    protected override string ShortName => "list";


    public override void OnWireReceived(IProtocol proto, SerializationCtx ctx, UnsafeReader stream, IRdWireableDispatchHelper dispatchHelper)
    {
      var header = stream.ReadLong();
      var opType = header & ((1 << versionedFlagShift) - 1);
      var version = header >> versionedFlagShift;

      var index = stream.ReadInt();


      var kind = (AddUpdateRemove) opType;
      var value = default(V);
      var isPut = kind is AddUpdateRemove.Add or AddUpdateRemove.Update;
      if (isPut)
        value = ReadValueDelegate(ctx, stream);

      var lifetime = dispatchHelper.Lifetime;
      var definition = value != null ? TryPreBindValue(lifetime, value, index, true) : null;

      dispatchHelper.Dispatch(() =>
      {
        ReceiveTrace?.Log($"list `{Location}` ({RdId}) :: {kind} :: index={index} :: version = {version}{(isPut ? " :: value = " + value.PrintToString() : "")}");

        if (version != myNextVersion)
        {
          definition?.Terminate();
          Assertion.Fail("Version conflict for {0} Expected version {1} received {2}. Are you modifying a list from two sides?",
            Location,
            myNextVersion,
            version);
        }


      myNextVersion++;


        switch (kind)
        {
          case AddUpdateRemove.Add:
          {
            if (index < 0)
            {
              TryGetBindDefinitions(lifetime)?.Add(definition);
              myList.Add(value);
            }
            else
            {
              TryGetBindDefinitions(lifetime)?.Insert(index, definition);
              myList.Insert(index, value);
            }

            break;
}
          case AddUpdateRemove.Update:
          {
            if (TryGetBindDefinitions(lifetime) is {} definitions)
            {
              definitions[index]?.Terminate();
              definitions[index] = definition;
            }

            myList[index] = value;
            break;
          }

          case AddUpdateRemove.Remove:
          {
            if (TryGetBindDefinitions(lifetime) is {} definitions)
            {
              definitions[index]?.Terminate();
              definitions.RemoveAt(index);
            }

            myList.RemoveAt(index);
            break;
          }

          default:
            throw new ArgumentOutOfRangeException(kind.ToString());
        }
      });
    }


    [CanBeNull]
    [ItemCanBeNull]
    private SynchronizedList<LifetimeDefinition> TryGetBindDefinitions(Lifetime lifetime)
    {
      var definitions = myBindDefinitions;
      return lifetime.IsAlive ? definitions : null;
    }

#nullable restore
    private LifetimeDefinition? TryPreBindValue(Lifetime lifetime, V? value, int index, bool bindAlso)
    {
      if (OptimizeNested || !value.IsBindable())
        return null;

      var definition = new LifetimeDefinition { Id = value };
      try
      {
        value.PreBindPolymorphic(definition.Lifetime, this, "["+index+"]"); //todo name will be not unique when you add elements in the middle of the list
        if (bindAlso)
          value.BindPolymorphic();

        lifetime.Definition.Attach(definition, true);
        return definition;
      }
      catch
      {
        definition.Terminate();
        throw;
      }
    }
#nullable disable


    #endregion


    #region Read delegation

    IEnumerator IEnumerable.GetEnumerator()
    {
      return myList.GetEnumerator();
    }

    public IEnumerator<V> GetEnumerator()
    {
      return myList.GetEnumerator();
    }

    public bool Contains(V item)
    {
      return myList.Contains(item);
    }

    public void CopyTo(V[] array, int arrayIndex)
    {
      myList.CopyTo(array, arrayIndex);
    }


    public int Count => myList.Count;

    public bool IsReadOnly => myList.IsReadOnly;


    public int IndexOf(V item)
    {
      return myList.IndexOf(item);
    }

    public ISource<ListEvent<V>> Change => myList.Change;

    #endregion


    #region Write delegation

    public void Insert(int index, V value)
    {
      using (UsingLocalChange())
        myList.Insert(index, value);
    }


    public void RemoveAt(int index)
    {
      using (UsingLocalChange())
        myList.RemoveAt(index);
    }


    public V this[int index]
    {
      get => myList[index];
      set
      {
        using (UsingLocalChange())
          myList[index] = value;
      }
    }


    public bool Remove(V item)
    {
      using (UsingLocalChange())
      {
        return myList.Remove(item);
      }
    }


    public void Add(V item)
    {
      using (UsingLocalChange())
      {
        myList.Add(item);
      }
    }


    public void Clear()
    {
      using (UsingLocalChange())
      {
        myList.Clear();
      }
    }

    #endregion


    public void Advise(Lifetime lifetime, Action<ListEvent<V>> handler)
    {
      if (IsBound) AssertThreading();

      using (UsingDebugInfo())
        myList.Advise(lifetime, handler);
    }

    public override RdBindableBase FindByRName(RName rName)
    {
      var rootName = rName.GetNonEmptyRoot();
      var localName = rootName.LocalName.ToString();
      if (!localName.StartsWith("[") || !localName.EndsWith("]"))
        return null;

      var stringIndex = localName.Substring(1, localName.Length - 2);
      if (!int.TryParse(stringIndex, out var index))
        return null;

      if (!(myList.ElementAtOrDefault(index) is RdBindableBase element))
        return null;

      if (rootName == rName)
        return element;

      return element.FindByRName(rName.DropNonEmptyRoot());
    }


    public override void Print(PrettyPrinter printer)
    {
      base.Print(printer);
      if (!printer.PrintContent) return;

      printer.Print(" [");
      if (Count > 0) printer.Println();

      using (printer.IndentCookie())
      {
        foreach (var v in this)
        {
          v.PrintEx(printer);
          printer.Println();
        }
      }
      printer.Println("]");
    }

  }

}

File: rd-kt/rd-framework/src/main/kotlin/com/jetbrains/rd/framework/impl/RdList.kt
package com.jetbrains.rd.framework.impl

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.framework.base.*
import com.jetbrains.rd.util.collections.SynchronizedList
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.lifetime.LifetimeDefinition
import com.jetbrains.rd.util.lifetime.isAlive
import com.jetbrains.rd.util.parseFromOrdinal
import com.jetbrains.rd.util.reactive.IMutableViewableList
import com.jetbrains.rd.util.reactive.IScheduler
import com.jetbrains.rd.util.reactive.IViewableList
import com.jetbrains.rd.util.reactive.ViewableList
import com.jetbrains.rd.util.string.*
import com.jetbrains.rd.util.trace


@Suppress("UNUSED_PARAMETER")
class RdList<V : Any> private constructor(val valSzr: ISerializer<V>, private val list: ViewableList<V>, private var nextVersion: Long = 1L)
    : RdReactiveBase(), IMutableViewableList<V> by list {

    companion object : ISerializer<RdList<*>> {
        private enum class Op {Add, Update, Remove} // update versionedFlagShift when changing

        fun<V:Any> read(ctx: SerializationCtx, buffer: AbstractBuffer, valSzr: ISerializer<V>): RdList<V> {
            return RdList(valSzr, ViewableList(), buffer.readLong()).withId(RdId.read(buffer))
        }

        override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): RdList<*> {
            return read(ctx, buffer, Polymorphic())
        }

        override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: RdList<*>) : Unit = value.run {
            buffer.writeLong(nextVersion)
            rdid.write(buffer)
        }

        const val versionedFlagShift = 2 // update when changing Op
    }

    override fun deepClone(): IRdBindable = RdList(valSzr).also { for (elem in list) { it.add(elem.deepClonePolymorphic()) } }

    var optimizeNested: Boolean = false
    private var bindDefinitions: MutableList<LifetimeDefinition?>? = null


    private fun logmsg(op: Op, version: Long, key: Int, value: V? = null) : String {
        return "list `$location` ($rdid) :: ${op.name} :: key = ${key.printToString()}"+
            (version > 0).condstr   { " :: version = $version" }  +
            (value != null).condstr { " :: value = ${value.printToString()}" }
    }

    override fun unbind() {
        super.unbind()
        bindDefinitions = null
    }

    override fun preInit(lifetime: Lifetime, proto: IProtocol) {
        super.preInit(lifetime, proto)

        if (!optimizeNested) {

            val definitions = SynchronizedList<LifetimeDefinition?>()

            for (index in 0 until size) {
                val item = this[index]
                if (item != null) {
                    item.identifyPolymorphic(proto.identity, proto.identity.next(rdid))
                    definitions.add(tryPreBindValue(lifetime, item, index, false))
                }
            }

            lifetime.executeIfAlive {
                assert(bindDefinitions == null)
                bindDefinitions = definitions
            }
        }

        proto.wire.advise(lifetime, this)
    }

    override fun init(lifetime: Lifetime, proto: IProtocol, ctx: SerializationCtx) {
        super.init(lifetime, proto, ctx)

        if (!optimizeNested) {
            change.advise(lifetime) {
                if (isLocalChange) {
                    val definitions = tryGetBindDefinitions(lifetime) ?: return@advise

                    if (it !is IViewableList.Event.Add)
                        definitions[it.index]?.terminate()

                    if (it is IViewableList.Event.Remove)
                        definitions.removeAt(it.index)

                    val value = it.newValueOpt
                    if (it !is IViewableList.Event.Remove) {
                        value.identifyPolymorphic(proto.identity, proto.identity.next(rdid))
                        definitions.add(it.index, tryPreBindValue(lifetime, value, it.index, false))
                    }
                }
            }
        }

        localChange { advise(lifetime) lambda@{
            if (!isLocalChange) return@lambda

            proto.wire.send(rdid) { buffer ->
                val op = when (it) {
                    is IViewableList.Event.Add ->    Op.Add
                    is IViewableList.Event.Update -> Op.Update
                    is IViewableList.Event.Remove -> Op.Remove
                }
                buffer.writeLong(op.ordinal.toLong() or (nextVersion++ shl versionedFlagShift))
                buffer.writeInt(it.index)

                it.newValueOpt?.let { valSzr.write(ctx, buffer, it) }

                logSend.trace { logmsg(op, nextVersion-1, it.index, it.newValueOpt) }
            }

            if (!optimizeNested)
                it.newValueOpt.bindPolymorphic()
        }}
    }

    override fun findByRName(rName: RName): RdBindableBase? {
        if (rName == RName.Empty) return this
        val rootName = rName.getNonEmptyRoot()
        val localName = rootName.localName
        if (!localName.startsWith('[') || !localName.endsWith(']'))
            return null
        
        val stringIndex = localName.removeSurrounding("[", "]")
        val index = stringIndex.toIntOrNull() 
            ?: return null

        val element = list.getOrNull(index) as? RdBindableBase
            ?: return null
        
        if (rootName == rName)
            return element
        
        return element.findByRName(rName.dropNonEmptyRoot())
    }

    override fun onWireReceived(proto: IProtocol, buffer: AbstractBuffer, ctx: SerializationCtx, dispatchHelper: IRdWireableDispatchHelper) {
        val header = buffer.readLong()
        val version = header shr versionedFlagShift
        val op = parseFromOrdinal<Op>((header and ((1 shl versionedFlagShift) - 1L)).toInt())
        val index = buffer.readInt()


        val value = if ((op == Op.Add || op == Op.Update)) valSzr.read(ctx, buffer) else null

        logReceived.trace { logmsg(op, version, index, value) }

        val lifetime = dispatchHelper.lifetime
        val definition = tryPreBindValue(lifetime, value, index, true)

        dispatchHelper.dispatch {
            if (version != nextVersion) {
                definition?.terminate()
                error("Version conflict for $location}. Expected version $nextVersion, received $version. Are you modifying a list from two sides?")
            }

            nextVersion++

            val definitions = tryGetBindDefinitions(lifetime)
            when (op) { // todo: better conflict resolution
                RdList.Companion.Op.Add -> {
                    if (index < 0) {
                        definitions?.add(definition)
                        list.add(value!!)
                    } else {
                        definitions?.add(index, definition)
                        list.add(index, value!!)
                    }
                }

                RdList.Companion.Op.Update -> {
                    if (definitions != null) {
                        definitions[index]?.terminate()
                        definitions[index] = definition
                    }
                    list[index] = value!!
                }

                RdList.Companion.Op.Remove -> {
                    definitions?.removeAt(index)?.terminate()
                    list.removeAt(index)
                }
            }
        }
    }

    private fun tryGetBindDefinitions(lifetime: Lifetime): MutableList<LifetimeDefinition?>?  {
        val definitions = bindDefinitions
        return if (lifetime.isAlive) definitions else null
    }

    private fun tryPreBindValue(lifetime: Lifetime, value: V?, index: Int, bindAlso: Boolean): LifetimeDefinition? {
        if (optimizeNested || value == null)
            return null

        val definition = LifetimeDefinition().apply { id = value }
        try {
            value.preBindPolymorphic(definition.lifetime, this, "[$index]")
            if (bindAlso)
                value.bindPolymorphic()

            (lifetime as LifetimeDefinition).attach(definition, true)

            return definition
        } catch (e: Throwable) {
            definition.terminate()
            throw e
        }
    }

    constructor(valSzr: ISerializer<V> = Polymorphic<V>()) : this(valSzr, ViewableList(SynchronizedList()))



    override fun print(printer: PrettyPrinter) {
        super.print(printer)
        printer.print(" [")
        if (!isEmpty()) printer.println()

        printer.indent {
            forEach {
                it.print(printer)
                printer.println()
            }
        }
        printer.print("]")
    }

    override fun iterator() = listIterator(0)
    override fun add(element: V) = localChange { list.add(element) }
    override fun add(index: Int, element: V) = localChange { list.add(index, element) }
    override fun addAll(index: Int, elements: Collection<V>): Boolean = localChange { list.addAll(index, elements) }
    override fun addAll(elements: Collection<V>): Boolean = localChange { list.addAll(elements) }
    override fun listIterator() = listIterator(0)
    override fun listIterator(index: Int): MutableListIterator<V> {
        val iterator = list.listIterator(index)
        return object : MutableListIterator<V> by iterator {
            override fun add(element: V) = localChange { iterator.add(element) }
            override fun remove() = localChange { iterator.remove() }
            override fun set(element: V) = localChange { iterator.set(element) }
        }
    }
    override fun remove(element: V): Boolean = localChange { list.remove(element) }
    override fun removeAll(elements: Collection<V>): Boolean = localChange { list.removeAll(elements) }
    override fun removeAt(index: Int): V = localChange { list.removeAt(index) }
    override fun retainAll(elements: Collection<V>): Boolean = localChange { list.retainAll(elements) }
    override fun set(index: Int, element: V): V = localChange { list.set(index, element) }
    override fun clear() = localChange { list.clear() }


    override fun advise(lifetime: Lifetime, handler: (IViewableList.Event<V>) -> Unit) {
        if (isBound) assertThreading()
        list.advise(lifetime, handler)
    }

    fun adviseOn(lifetime: Lifetime, scheduler: IScheduler, handler: (IViewableList.Event<V>) -> Unit) {
        if (isBound) assertThreading()
        list.advise(lifetime) { e -> scheduler.invokeOrQueue { handler(e) }}
    }


}
